GHzFPGA.run: Enable ADC Average Mode

jwenner opened the pull request:
Fixes #100.

In FPGA server/run (and hence run_sequence), if ADC in average mode so that no demod channel specified, raised 'channel format not understood' exception. Now add elif case in getTimingData so if ADC is present in channel name but not :: (for demodulation), can still return data. Move calls to channel to if statement for if channel defined. Amend return format type for run_sequence to also allow '*3i' for average mode data.

Note that in going between demod and average modes (although in different scans & calls to run_sequence), it seems to work fine (although it's possible I might not be seeing/knowing what @JulianSKelly was referring to).

DanielSank commented:
Why is the return type `*3i` for average mode? Shouldn't it be `*2i`?

jwenner commented:
@DanielSank, I agree it should be ```*2i```. However, at least fpgaTest assumes that it be ```*3i```, so I would propose that we stick with this for backwards compatibility.

Actually, before average mode was removed, the type tag for average mode was ```?```.

DanielSank commented:
> @DanielSank, I agree it should be *2i. However, at least fpgaTest assumes that it be *3i, so I would propose that we stick with this for backwards compatibility.

I think we should fix it. What's the extra dimension for?

jwenner commented:
@DanielSank, the dimensions from run_sequence appear to be as follows:
   * demod: [demodChannel, trialNumber, data?, data?]
   * average: [demodChannel, IOrQ, data]

I would propose that we instead make the average mode data 4-dimensional of the same form as demod (i.e., [demodChannel, trialNumber, IOrQ, data]), with the length=1 in the demodChannel and trialNumber dimensions. The reason would be so that we could (maybe?) take average mode & demod mode data at the same time on different ADCs (e.g., measure qubit state &, say, resonator reflection at the same time).

Would changing any of this (in any direction) require changes in the qubit sequencer server?

Thanks.

DanielSank commented:
> I would propose that we instead make the average mode data 4-dimensional of the same form as demod (i.e., [demodChannel, trialNumber, IOrQ, data])

The demodulator mode data indices are `(channel, stat, demod, I/Q)`. I'm not sure what you mean by `data` for the final index. I also don't understand what it means to have `channel` and `trialNumber` indices on average mode data, as the ADC board does not support separate traces for different trials. In other words, average mode doesn't have "stats" in the same was as demodulator mode.

jwenner commented:
@DanielSank, happy now?

PS: Upon thinking about it, I believe the third dimension was for if we had multiple ADCs present, from which we were reading back average mode data simultaneously.

ejeffrey commented:
As far as I know, we can't currently support having some boards in average mode and some in IQ mode without major changes, and even if we did, we couldn't return the data in a 4D array since the channels would not all be the same length.  I don't think trying to be future compatible in this way is going to help us any.

Having a third index for 'board number' in the case of multiple boards is fine, but in any case the axes of the return data should be documented!  If it said "the return in average mode is a 3D array of (board#, time, I/Q) data" we wouldn't be having this conversation. Same for the IQ mode.

jwenner commented:
@DanielSank, @ejeffrey, could you please now take a look at this ghz_fpga_server update?

* Gone back to the 3D array to handle the multiple ADC boards case. 
* Updated the run_sequence docstring describing the return data format so it's no longer completely out of date. 
* Cleaned up some of the comments in the part of run I've dealt with (including renaming one variable which was multiply defined w/ a comment saying why OK but bad).

jwenner commented:
@DanielSank, if/when you approve this and I merge it back in, can we also close martinisgroup/pyle#103 as well (considering the updates to the fpga.run_sequence docstring)? Thanks.

DanielSank commented:
@jwenner This looks great. Thanks for moving the comments around so that they make more sense.

I would like to fix up the documentation of the `Run Sequence` setting, as the notation used there for describing the data format seems to be in Jim Wenner Markdown<sup>TM</sup>. If you don't mind I'll add a commit to this branch.

DanielSank commented:
LGTM.

Jim, when rebasing please try to keep the changes in content separate from the documentation and white space.

This is great, thanks.

zchen088 commented:
>Jim, when rebasing please try to keep the changes in content separate from the documentation and white space.

This is a thing we're supposed to do when rebasing? I thought the whitespace thing was only for the initial reviewing, and I've been squashing all of my pull requests down to one commit. I also wasn't aware that documentation needs to be in a separate commit. Can this get clarified in the contributing doc?

Also, can we get an explanation in the contributing guide as to why this needs to be the case?

jwenner commented:
@DanielSank, most of your commit looks good. However, there was an error in your revision to the docstring for average mode (order of indices) that I've now fixed, and I also added a sentence to clarify what the retrigger index is. Can you take a quick look and tell me if this is OK?

For the documentation commit, can I mark martinisgroup/pyle#103 closed?

Thanks.

DanielSank commented:
> However, there was an error in your revision to the docstring for average mode (order of indices) that I've now fixe

@jwenner Thanks for fixing that mistake.

> For the documentation commit, can I mark martinisgroup/pyle#103 closed?

Yes, as soon as we merge.

> Can you take a quick look and tell me if this is OK?

Yeah, looks great.
